### DISCUSSION TITLE
在 Golang 中如何做国际化？

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#407] [Announcements]
# 在 Golang 中如何做国际化？

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/2019/12/11/golang-i18n/

**Comment by xuxinx** [comment]
hello, 我最近打算使用https://golang.org/x/text 来做一个项目的翻译。调研过程中有个疑问，输出的json文件中的id和message这两个字段内容总是一致的，我比较疑惑这样的意义在哪里？（要向translators解释这个东西）
```json
        {
            "id": "Welcome!",
            "message": "Welcome!",
            "translation": "Welcome ttt!"
        }
```

**Reply by Xuanwo** [reply]
我不太清楚这个“输出的 JSON”是从哪里来的，不过其本身的含义还是比较明确的，换一个 naming 可能会更好理解：

```json
{
  "id": "Welcome!",
  "en_US": "Welcome!",
  "zh_Hans": "欢迎！"
}
```

id 是待翻译字符串的唯一 id，用于在各种地方引用，这里跟 message 一致应该只是为了生成方便

**Reply by xuxinx** [reply]
> “输出的 JSON”是从哪里来的

通过[gotext](https://pkg.go.dev/golang.org/x/text/cmd/gotext)从源码中获得的待翻译文件。例子 https://github.com/golang/text/blob/master/cmd/gotext/examples/extract/locales/en-US/out.gotext.json

**Reply by Xuanwo** [reply]
这个我就不太了解了，感觉还涉及到了 i10n :rofl:
